### ISSUE TITLE
Make storage pluggable

### ISSUE THREAD

**Issue by thunderbird** [spring-cloud/spring-cloud-dataflow#934]
Currently one is forced to use RDBMS for storing streams, tasks, etc. This doesn't fully follow the rest of the spring philosophy... It would be good if we can make this customisable so that we can use alternative types of databases for doing this (like cassandra, rial, mongo, redis, etc).

Analytics also requires Redis. Making this customisable as well would make this project sing.

**Comment by sabbyanandan** [comment]
Hi, @thunderbird:

> Currently one is forced to use RDBMS for storing streams, tasks, etc

Building upon [Spring Batch](http://projects.spring.io/spring-batch/) and [Spring Cloud Task](http://cloud.spring.io/spring-cloud-task/) projects, Spring Cloud Data Flow (SCDF) supports short-lived, offline, and batch-style data processing use-cases. Spring Batch in particular defaults to an RDBMS based `JobRespoitory`, which is essential for use-cases that require transactions, restartability and to prevent duplicate concurrent launches etc. 

Given that we were already using an RDBMS for Task/Batch in SCDF, we have defaulted the persistence of stream/task definitions in the same datasource. That said, [StreamDefinitionRepository](https://github.com/spring-cloud/spring-cloud-dataflow/blob/master/spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/repository/StreamDefinitionRepository.java) and [TaskDefinitionRepository](https://github.com/spring-cloud/spring-cloud-dataflow/blob/master/spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/repository/TaskDefinitionRepository.java) are interfaces, so you could extend it to non-relational datasources as needed. 

> Analytics also requires Redis. Making this customisable as well would make this project sing.

Likewise, we have Redis implementation of [`AggregateCounterRepository`](https://github.com/spring-projects/spring-analytics/blob/master/src/main/java/org/springframework/analytics/metrics/redis/RedisAggregateCounterRepository.java) and [`FieldValueCounterRepository`](https://github.com/spring-projects/spring-analytics/blob/master/src/main/java/org/springframework/analytics/metrics/redis/RedisFieldValueCounterRepository.java) and again, this can be extended to other stores. 

Until we add the new implementations, we encourage users for contributions as they see fit. :)

**Comment by thunderbird** [comment]
I'll take a look. Thanks, more than happy to contribute... I just need to get permission internally first :)

**Comment by mminella** [comment]
There are some things that make this difficult.  We probably should have a conversation here before the related PR is merged.  Specific issues I see with this:
1. **Domain Objects -** The way the domain objects are defined right now are not friendly to multiple data store options.  To use something like Spring Data Redis or Spring Data Mongo, you'd need to add mapping annotations to the domain objects which may be data store specific.  The alternative would be to implement these mappings at a much lower level.  Another issue with the domain objects is that they assume a simple key (`long` in the task repository case).  Data stores like Casandra don't support keys like that.
2.  **This doesn't fully follow the rest of the spring philosophy -** I'd have to disagree.  Spring Batch requires a RDBMS for storing the job repository metadata.  Spring Cloud Data Flow isn't really a framework like the rest of the Spring ecosystem, it's a server implementation.  It's a fully baked application.  It is open source so it can be extended, but there are design decisions that have been made that impact those extensibility options.  The goal has been to provide a server implementation that is as approachable to the masses and extensible/configurable in every reasonable way we can while still making sound architectural decisions.
3.  **Dependencies on other projects -** The changing of some of these tables is not as simple as just changing the server implementation.  Spring Cloud Task applications share the tables with Spring Cloud Data Flow.  In order for this to work, you not only need a server that supports the new data store, but your apps need to as well.  Same goes for the Spring Batch job repository tables.

**Comment by tzolov** [comment]
Here is my take on this story:
  
As @mminella pointed out, `SCDF` is not (only) a framework but a runtime application with an internal, persistable state!  So it is discourageable to access it by any other means but SCDF itself. (e.g. no external, direct (No)SQL access to the internal state).

The question is:
> What does SCDF gains of maintaining multi data-storage support for its INTERNAL state? 

* This "capability" in fact limits the extensibility of the internal state model!
* Perhaps SCDF should focus on selecting a storage that best serves its primary purpose of orchestrating complex data pipelines. For example if the persistent storage provides write-ahead-log (WAL) capabilities, this would allow implementing reliable recovery guaranties for SCDF streams as well as help implementing provenance features...
* The analytics seems to have the opposite requirements. It might be beneficial to let external, advanced analytics and visualization tools access the SCDF analytics data. That means SCDF should invest in supporting different formats (time-series for example) and storage systems.
* Batch's dependencies on RDBMS(Tx) should be dealt in the scope for Batch/Task but not spread across the entire SCDF

In short, SCDF should be more opinionated about its internal state storage and be more extensible with regards to analytic formats and storages.

**Comment by sabbyanandan** [comment]
Thanks for the feedback, @tzolov. For this to happen, we will first have to split Task/Batch schema from SCDF's, and have them point to different DB instances. We will revisit the requirement in the context of 1.4 and onwards. 

I will close this for now.
